test/core/iomgr/threadpool_test.cc

```diff
@@ -0,0 +1,173 @@
+    pool->Add(functor);
+  }
+  grpc_core::Delete(pool);
+  GPR_ASSERT(functor->count() == THREAD_SMALL_ITERATION);
+  grpc_core::Delete(functor);
+  gpr_log(GPR_DEBUG, "Done.");
+}
+
+// Thread that adds closures to pool
+class WorkThread {
+ public:
+  WorkThread(grpc_core::ThreadPool* pool, SimpleFunctorForAdd* cb, int num_add)
+      : num_add_(num_add), cb_(cb), pool_(pool) {
+    thd_ = grpc_core::Thread(
+        "thread_pool_test_add_thd",
+        [](void* th) { static_cast<WorkThread*>(th)->Run(); }, this);
+  }
+  ~WorkThread() {}
+
+  void Start() { thd_.Start(); }
+  void Join() { thd_.Join(); }
+
+ private:
+  void Run() {
+    for (int i = 0; i < num_add_; ++i) {
+      pool_->Add(cb_);
+    }
+  }
+
+  int num_add_;
+  SimpleFunctorForAdd* cb_;
+  grpc_core::ThreadPool* pool_;
+  grpc_core::Thread thd_;
+};
+
+static void test_multi_add(void) {
+  gpr_log(GPR_INFO, "test_multi_add");
+  const int num_work_thds = 10;
+  grpc_core::ThreadPool* pool = grpc_core::New<grpc_core::ThreadPool>(
+      LARGE_THREAD_POOL_SIZE, "test_multi_add");
+  SimpleFunctorForAdd* functor = grpc_core::New<SimpleFunctorForAdd>();
+  WorkThread** work_thds = static_cast<WorkThread**>(
+      gpr_zalloc(sizeof(WorkThread*) * num_work_thds));
+  gpr_log(GPR_DEBUG, "Fork threads for adding...");
+  for (int i = 0; i < num_work_thds; ++i) {
+    work_thds[i] =
+        grpc_core::New<WorkThread>(pool, functor, THREAD_LARGE_ITERATION);
+    work_thds[i]->Start();
+  }
+  // Wait for all threads finish
+  gpr_log(GPR_DEBUG, "Waiting for all work threads finish...");
+  for (int i = 0; i < num_work_thds; ++i) {
+    work_thds[i]->Join();
+    grpc_core::Delete(work_thds[i]);
+  }
+  gpr_free(work_thds);
+  gpr_log(GPR_DEBUG, "Done.");
+  gpr_log(GPR_DEBUG, "Waiting for all closures finish...");
+  // Destructor of thread pool will wait for all closures to finish
+  grpc_core::Delete(pool);
+  GPR_ASSERT(functor->count() == THREAD_LARGE_ITERATION * num_work_thds);
+  grpc_core::Delete(functor);
+  gpr_log(GPR_DEBUG, "Done.");
+}
+
+static void test_one_thread_FIFO(void) {
+  gpr_log(GPR_INFO, "test_one_thread_FIFO");
+  grpc_core::ThreadPool* pool =
+      grpc_core::New<grpc_core::ThreadPool>(1, "test_one_thread_FIFO");
+  SimpleFunctorForAdd* functor = grpc_core::New<SimpleFunctorForAdd>();
+  SimpleFunctorCheckForAdd** check_functors =
+      static_cast<SimpleFunctorCheckForAdd**>(gpr_zalloc(
+          sizeof(SimpleFunctorCheckForAdd*) * THREAD_SMALL_ITERATION));
+  for (int i = 0; i < THREAD_SMALL_ITERATION; ++i) {
+    pool->Add(functor);
+    check_functors[i] =
+        grpc_core::New<SimpleFunctorCheckForAdd>(functor, i + 1);
+    pool->Add(check_functors[i]);
+  }
+  // Destructor of pool will wait until all closures finished.
+  grpc_core::Delete(pool);
+  grpc_core::Delete(functor);
+  for (int i = 0; i < THREAD_SMALL_ITERATION; ++i) {
+    grpc_core::Delete(check_functors[i]);
+  }
+  gpr_free(check_functors);
+  gpr_log(GPR_DEBUG, "Done.");
+}
+
```

Comment:
I think we also need test cases covering options constructor and num_threads == 0.